### PR TITLE
Add domain option, GNIX_MR_DISABLE_CACHING

### LINKED ISSUE #832: Add option to disable memory registration cache
Useful for clients that want to implement their own cache.

### PR DESCRIPTION
**Pull request by jswaro**
New domain option allows the memory registration cache to be
disabled on request.

Signed-off-by: James Swaro jswaro@cray.com

closes #832 
@sungeunchoi @hppritcha

### CODE REVIEW

**Comment by sungeunchoi** [comment]
To me, it feels confusing to have GNI_MR_CACHE_UDREG and GNI_MR_DISABLE_CACHING.  I would prefer to have a single domain op variable called, say, GNI_MR_CACHE.  Possible values would be: none, udreg, and default/internal (and various capitalization of these words).

**Comment by sungeunchoi** [comment]
Does udreg have a way to disable lazy deregistration at runtime?  If so, it should be hooked up to the appropriate domain op for disabling.

File: prov/gni/src/gnix_mr.c [hunk 1]
```diff
@@ -486,6 +486,9 @@ static int __udreg_init(struct gnix_fid_domain *domain)
 				"Could not access udreg application cache, urc=%d",
 				urc);
 	}
+
+	domain->mr_is_init = 1;
+
 	return 0;
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_mr.c]: Sorry I did not catch this before...  to be consistent, this should return FI_SUCCESS.  This happens in a few other places too.

**Comment by jswaro** [comment]
I would be comfortable with exposing the type enumerations but I'd rather stay away from string comparisons. Suggestions?

--Jim

**Comment by jswaro** [comment]
> Does udreg have a way to disable lazy deregistration at runtime? If so, it should be hooked up to the appropriate domain op for disabling.

When you create the cache, you can provide lazy deregistration as a flag. We can create a tunable domain op that sets or unsets this flag prior to creating the cache.

File: prov/gni/src/gnix_mr.c (hunk 1 above)
Comment (jswaro) [inline prov/gni/src/gnix_mr.c]: fixed in the next push

**Comment by sungeunchoi** [comment]
Can we reuse the lazy dereg flag we have now?

**Comment by jswaro** [comment]
> Can we reuse the lazy dereg flag we have now?

I don't see why not. Would you mind opening an issue for that?

**Comment by sungeunchoi** [comment]
Done.  See #845

**Comment by sungeunchoi** [comment]
👍 Thx!

**Pull request merged**